connectivity/tests/encryption.go

```diff
@@ -0,0 +1,178 @@
+// then inspecting tcpdump captures from the client pod's node.
+func PodToPodEncryption() check.Scenario {
+	return &podToPodEncryption{}
+}
+
+type podToPodEncryption struct{}
+
+func (s *podToPodEncryption) Name() string {
+	return "pod-to-pod-encryption"
+}
+
+func (s *podToPodEncryption) Run(ctx context.Context, t *check.Test) {
+	client := t.Context().RandomClientPod()
+
+	var server check.Pod
+	for _, pod := range t.Context().EchoPods() {
+		// Make sure that the server pod is on another node than client
+		if pod.Pod.Status.HostIP != client.Pod.Status.HostIP {
+			server = pod
+			break
+		}
+	}
+
+	bgCtx, bgCancel := context.WithCancel(ctx)
+
+	// clientHost is a pod running on the same node as the client pod, just in
+	// the host netns.
+	clientHost := t.Context().HostNetNSPodsByNode()[client.Pod.Spec.NodeName]
+
+	// Determine on which netdev iface to capture pkts. In the case of tunneling,
+	// we don't expect to see unencrypted pkts on a corresponding tunneling iface,
+	// so the choice is obvious. In the native routing mode, we run
+	// "ip route get $SERVER_POD_IP" from the client pod's node.
+	iface := ""
+	tunnelFeat, ok := t.Context().Feature(check.FeatureTunnel)
+	if ok && tunnelFeat.Enabled {
+		iface = "cilium_" + tunnelFeat.Mode // E.g. cilium_vxlan
+	} else {
+		cmd := []string{"/bin/sh", "-c",
+			fmt.Sprintf("ip -o r g %s from %s | grep -oP '(?<=dev )[^ ]+'",
+				server.Pod.Status.PodIP, client.Pod.Status.PodIP)}
+		t.Debugf("Running %s", strings.Join(cmd, " "))
+		dev, err := clientHost.K8sClient.ExecInPod(ctx, clientHost.Pod.Namespace,
+			clientHost.Pod.Name, "", cmd)
+		if err != nil {
+			t.Fatalf("Failed to get IP route: %s", err)
+		}
+		iface = strings.TrimRight(dev.String(), "\n\r")
+	}
+
+	t.Debugf("Detected %s iface for communication among client and server nodes",
+		iface)
+
+	bgStdout := &safeBuffer{}
+	bgStderr := &safeBuffer{}
+	bgExited := make(chan struct{})
+	// Start kubectl exec in bg (=goroutine)
+	go func() {
+		// Run tcpdump with -w instead of directly printing captured pkts. This
+		// is to avoid a race after sending ^C (triggered by bgCancel()) which
+		// might terminate the tcpdump process before it gets a chance to dump
+		// its captures.
+		cmd := []string{
+			"tcpdump", "-i", iface, "--immediate-mode", "-w", "/tmp/foo.pcap",
+			// Capture pod egress traffic.
+			// Unfortunately, we cannot use "host %s and host %s" filter here,
+			// as IPsec recirculates replies to the iface netdev, which would
+			// make tcpdump to capture the pkts (false positive).
+			fmt.Sprintf("src host %s and dst host %s", client.Pod.Status.PodIP, server.Pod.Status.PodIP),
+			// Only one pkt is enough, as we don't expect any unencrypted pkt
+			// to be captured
+			"-c", "1"}
+		t.Debugf("Running in bg: %s", strings.Join(cmd, " "))
+		err := clientHost.K8sClient.ExecInPodWithWriters(bgCtx, clientHost.Pod.Namespace,
+			clientHost.Pod.Name, "", cmd, bgStdout, bgStderr)
```

Comment:
`bgStderr` is only written to but never read. Is this intended? Otherwise we could just use

```suggestion
			clientHost.Pod.Name, "", cmd, bgStdout, io.Discard)
```

Reply:
Makes sense. But in that case we should probably check (or at least debug log) `bgStderr` contents somewhere after executing tcpdump?

Reply:
My reading of the `Exec*` code path is that the stderr is being pushed into the stdout later on.

Reply:
Yes, so why would we need to provide `bgStderr` then? I think the stderr->stdout merge happens inside client-go if `PodExecOptions.TTY` is set, so I don't think we need to in provide an `io.Writer` for `stderr`?

In any case, with #1276 the use of `PodExecOptions.TTY` will hopefully go away soon and we'd want to populate an `io.Writer` for `stderr` (and check it at the `ExecInPod*` callsite). That being said, I'm ok to leave that as is and will clean up as part of #1276 after rebase.

Thanks!

Reply:
> That being said, I'm ok to leave that as is and will clean up as part of https://github.com/cilium/cilium-cli/pull/1276 after rebase.

Yeah, let's do that!